### ISSUE TITLE
tensort knowledges

### ISSUE THREAD

**Issue by WarriorHanamy** [WarriorHanamy/realworld_modules#14]
By default, TensorRT engines are compatible only with the version of TensorRT used to build them. With appropriate build-time configuration, you can build engines that are compatible with later TensorRT versions.

For example, **engines built with TensorRT 8 work with TensorRT 9 and TensorRT 10 runtimes.** However, engines built with newer versions (TensorRT 9 or 10) do not work with older runtimes (TensorRT 8).

**Comment by WarriorHanamy** [comment]
**By default, TensorRT engines are only compatible with the type of device where they were built.** With build-time configuration, engines that are compatible with other types of devices can be built. Currently, hardware compatibility is not supported on NVIDIA DRIVE OS or JetPack.

**Comment by WarriorHanamy** [comment]
So not “build or install” as an either/or:

usually install TensorRT
optionally build TensorRT OSS addons/components

That is the pattern I’d recommend on Jetson.

**Comment by WarriorHanamy** [comment]
n traditional software development, a cross-compiler like GCC translates high-level code (C++) into machine code for a different CPU architecture (e.g., x86 to ARM). This is possible because the instruction set architectures are well-defined and abstracted.

In contrast, a TensorRT engine is the result of a runtime optimization process, not just a compilation step. The builder makes real-time decisions based on the hardware it can query right now.

**Comment by WarriorHanamy** [comment]
10.16.1 tensorrt

ONNX Model Support

The ONNX-TensorRT parser supports:

ONNX version: [1.20.0](https://github.com/onnx/onnx/releases/tag/v1.20.0)

Opset support: Up to opset 25

Backward compatibility: Official support is provided for opset 9 and above

For more information, refer to the [TensorRT Support Matrix](https://docs.nvidia.com/deeplearning/tensorrt/latest/getting-started/support-matrix.html#support-matrix).

**Comment by WarriorHanamy** [comment]
Upgrading TensorRT
Upgrading TensorRT to the latest version is only supported when the currently installed TensorRT version is equal to or newer than the last two public GA releases.

If you want to upgrade from an unsupported version, you should incrementally upgrade until you reach the latest version of TensorRT or uninstall and reinstall the newest version. If you have an EA version of TensorRT installed, you should first upgrade to the corresponding GA version.

Upgrading From TensorRT 10.x.x To TensorRT 10.16.x
When upgrading from TensorRT 10.x.x to TensorRT 10.16.x, ensure you are familiar with the following.

Using pip

Use the following pip commands to upgrade a Python Package Index (pip) installation:

python3 -m pip cache remove "tensorrt*"
python3 -m pip install --upgrade tensorrt tensorrt-lean tensorrt-dispatch

**Comment by WarriorHanamy** [comment]
<img width="1197" height="1173" alt="Image" src="https://github.com/user-attachments/assets/930cd44c-7437-4913-8f9b-256915004ab4" />

**Comment by WarriorHanamy** [comment]
TensorRT’s primary means of importing a trained model from a framework is the [ONNX](https://onnx.ai/) interchange format. TensorRT ships with an ONNX parser library to assist in importing models. Where possible, the parser is backward compatible to opset 9. The [ONNX Model Opset Version Converter](https://github.com/onnx/onnx/blob/master/docs/VersionConverter.md) can assist in resolving incompatibilities.

The [GitHub version](https://github.com/onnx/onnx-tensorrt/) can support later opsets than the version shipped with TensorRT. Refer to the [ONNX-TensorRT operator support matrix](https://github.com/onnx/onnx-tensorrt/blob/master/docs/operators.md) for the latest information on the supported opset and operators. **For TensorRT deployment, we recommend exporting to the latest available ONNX opset.**

**Comment by WarriorHanamy** [comment]
While using FP16 and TF32 precisions is relatively straightforward, **working with INT8 adds additional complexity.** For more information, refer to the [Working with Quantized Types](https://docs.nvidia.com/deeplearning/tensorrt/10.16.1/inference-library/work-quantized-types.html#work-with-quantized-types) section.

**Comment by WarriorHanamy** [comment]
**We should be cautious about dynamics shapes**
**Dynamic shapes**	Not handled. The network flag EXPLICIT_BATCH is set, but no profile is created. If the ONNX model has dynamic dimensions, building will likely fail or produce a suboptimal engine.	Can define multiple optimization profiles to handle dynamic input shapes efficiently.

**Comment by WarriorHanamy** [comment]
Your build_engine is a minimal, embedded‑style builder—perfect for **quick prototyping** or integrating into a Python application when you only need basic FP16/static‑shape conversion.

trtexec is a production‑grade benchmarking and deployment tool that exposes all TensorRT knobs, handles complex input shapes, and validates the final engine

**Comment by WarriorHanamy** [comment]
Build a TensorRT engine from ONNX using the [trtexec](https://github.com/NVIDIA/TensorRT/tree/main/samples/trtexec) tool.

trtexec can generate a TensorRT engine from an ONNX model that can then be deployed using the **TensorRT runtime API.** It leverages the [TensorRT ONNX parser](https://github.com/onnx/onnx-tensorrt) to load the ONNX model into a TensorRT network graph and the [**TensorRT Builder API**](https://docs.nvidia.com/deeplearning/tensorrt/10.16.1/inference-library/c-api-docs.html#build-engine-c) to generate an optimized engine.

**Comment by WarriorHanamy** [comment]
There are four main options for converting a model with TensorRT:

Using Torch-TensorRT

Automatic ONNX conversion from .onnx files

Using the GUI-based tool [Nsight Deep Learning Designer](https://developer.nvidia.com/nsight-dl-designer)

Manually constructing a network using the TensorRT API (either in C++ or Python)

**Comment by WarriorHanamy** [comment]
[ONNX](https://github.com/onnx/onnx/blob/main/docs/IR.md) is a framework-agnostic model format that can be exported from most major frameworks, including TensorFlow and PyTorch. **TensorRT provides a library for directly converting ONNX into a TensorRT engine through the [ONNX-TRT parser](https://github.com/onnx/onnx-tensorrt).**

**Comment by WarriorHanamy** [comment]
https://docs.nvidia.com/deeplearning/frameworks/container-release-notes/

**Comment by WarriorHanamy** [comment]
1. I/O tensors

These are the things your application owns and changes every inference:

input images
input feature tensors
output logits
output boxes / scores / embeddings

These need buffers, addresses, copies, and synchronization. That is exactly what your HostDeviceMem, bindings, cudaMemcpyAsync, and cudaStreamSynchronize are for.

2. Engine internals

These include:

layer weights
biases
fused constants
execution plan details
kernel choices
workspace requirements

These are part of the TensorRT engine/runtime internals. You do not allocate them in this loop, and you do not put them into bindings.

**Comment by WarriorHanamy** [comment]
NVIDIA documents that trtexec is included with TensorRT. In current Linux installs, NVIDIA says trtexec is now installed to /usr/bin and added to PATH; older installs commonly used /usr/src/tensorrt/bin/trtexec, and NVIDIA examples for Jetson still reference that path.
